client/tx/utils.go

```diff
@@ -0,0 +1,173 @@
+package tx
+
+import (
+	"encoding/hex"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+// SearchTxs performs a search for transactions for a given set of tags via
+// Tendermint RPC. It returns a slice of Info object containing txs and metadata.
+// An error is returned if the query fails.
+func SearchTxs(cliCtx context.CLIContext, cdc *codec.Codec, tags []string, page, limit int) ([]sdk.TxResponse, error) {
+	if len(tags) == 0 {
+		return nil, errors.New("must declare at least one tag to search")
+	}
+
+	if page <= 0 {
+		return nil, errors.New("page must greater than 0")
+	}
+
+	if limit <= 0 {
+		return nil, errors.New("limit must greater than 0")
+	}
+
+	// XXX: implement ANY
+	query := strings.Join(tags, " AND ")
+
+	node, err := cliCtx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	prove := !cliCtx.TrustNode
+
+	resTxs, err := node.TxSearch(query, prove, page, limit)
+	if err != nil {
+		return nil, err
+	}
+
+	if prove {
+		for _, tx := range resTxs.Txs {
+			err := ValidateTxResult(cliCtx, tx)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	resBlocks, err := getBlocksForTxResults(cliCtx, resTxs.Txs)
+	if err != nil {
+		return nil, err
+	}
+
+	txs, err := formatTxResults(cdc, resTxs.Txs, resBlocks)
+	if err != nil {
+		return nil, err
+	}
+
+	return txs, nil
+}
+
+// formatTxResults parses the indexed txs into a slice of TxResponse objects.
+func formatTxResults(cdc *codec.Codec, resTxs []*ctypes.ResultTx, resBlocks map[int64]*ctypes.ResultBlock) ([]sdk.TxResponse, error) {
+	var err error
+	out := make([]sdk.TxResponse, len(resTxs))
+	for i := range resTxs {
+		out[i], err = formatTxResult(cdc, resTxs[i], resBlocks[resTxs[i].Height])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+// ValidateTxResult performs transaction verification.
+func ValidateTxResult(cliCtx context.CLIContext, resTx *ctypes.ResultTx) error {
+	if !cliCtx.TrustNode {
+		check, err := cliCtx.Verify(resTx.Height)
+		if err != nil {
+			return err
+		}
+		err = resTx.Proof.Validate(check.Header.DataHash)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getBlocksForTxResults(cliCtx context.CLIContext, resTxs []*ctypes.ResultTx) (map[int64]*ctypes.ResultBlock, error) {
```

Comment:
can we use uint for the blocks too ?
```suggestion
func getBlocksForTxResults(cliCtx context.CLIContext, resTxs []*ctypes.ResultTx) (map[uint64]*ctypes.ResultBlock, error) {
```

Reply:
There would be no point. Tendermint uses `int64` for block heights, so we'd end up doing a pointless casting. It's safe here to use int64 as there is no user input or other means for invalid values.